### PR TITLE
247 Promote deployment

### LINKED ISSUE #247: Promote a deployment
### Description of problem

As an operator, I want to be able to promote an environment after a deployment has succeeded. In order to do so, I need a way to indicate that my deployment has succeeded. 

[config] ========> [kapp] =======> [promote]

### Proposed solution

- [x] Implement `observedCompletion` and `observedMatches` in the `ClusterDeploymentTemplate` as per https://github.com/sclevine/cartographer/blob/patch-1/rfc/rfc-0013-delivery.md
- [x] Allow `ClusterDeploymentTemplate` to be used as a `source` to downstream resources

### Example

Fetches Kubernetes configuration from a git repository
Deploys to a staging cluster using kapp controller
Promotes to production via pull request if integration tests pass

Deliverable

```yaml
apiVersion: carto.run/v1alpha1
kind: Deliverable
metadata:
  name: my-app
  labels:
    app.tanzu.vmware.com/deliverable-type: web
spec:
  config:
    git:
      url: https://github.com/ekcasey/hello-world-ops
      ref:
        branch: staging
```

ClusterDelivery

```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterDelivery
metadata:
  name: delivery
spec:
  selector:
    app.tanzu.vmware.com/deliverable-type: web
  resources:
    - name: config-provider
      templateRef:
        kind: ClusterSourceTemplate
        name: git-repository
    - name: deployer
      templateRef:
        kind: ClusterDeployTemplate
        name: app-deploy
      source:
        resource: config-provider
    - name: promoter
      templateRef:
        kind: ClusterTemplate
        name: git-merge
      sources:
      - resource: deployer
```

ClusterSourceTemplate

```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterSourceTemplate
metadata:
  name: source
spec:
  urlPath: .status.artifact.url
  revisionPath: .status.artifact.revision

  template:
    apiVersion: source.toolkit.fluxcd.io/v1beta1
    kind: GitRepository
    metadata:
      name: $(deliverable.metadata.name)$
    spec:
      interval: 1m
      url: $(deliverable.spec.source.git.url)$
      ref: $(deliverable.spec.source.git.ref)$
      gitImplementation: libgit2
      ignore: ""
```

ClusterDeploymentTemplate

```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterDeploymentTemplate
metadata:
  name: app-deploy
spec:
  observedCompletion: # uses observedGeneration
    succeeded: {key: 'status.conditions.#(type=="ReconcileSucceeded").status', value: "True"}

  template:
    apiVersion: kappctrl.k14s.io/v1alpha1
    kind: App
    metadata:
      name: $(deliverable.metadata.name)$
    spec:
      serviceAccountName: default
      fetch:
       - http:
          url: $(source.url)$
      template:
        - ytt: {}
      deploy:
        - kapp: {}
```

ClusterTemplate - Git Merge

```yaml
apiVersion: carto.run/v1alpha1
kind: ClusterTemplate
metadata:
  name: git-merge
spec:
  template:
    apiVersion: carto.run/v1alpha1
    kind: Pipeline
    metadata:
      generateName: $(workload.name)$-promotion-
    spec:
      inputs:
        url: $(source.url)$
        revision: $(source.revision)$
        pipeline: test

      runTemplateRef:
        name: default-tekton-template
```

PipelineRun - Tests

```yaml
apiVersion: carto.run/v1alpha1
kind: RunTemplate
metadata:
  name: default-tekton-source-provider
spec:
  completion:
    succeeded: {key: 'status.conditions.#(type=="Succeeded").status', value: "True"}
    failed:    {key: 'status.conditions.#(type=="Succeeded").status', value: "False"}

  template:
    apiVersion: tekton.dev/v1beta1
    kind: PipelineRun
    metadata:
      generateName: $(pipeline.metadata.name)$-
    spec:
      pipelineRef: 
        name: $(pipeline.spec.inputs.pipelineName)$

      workspaces:
        - name: source
          volumeClaimTemplate:
            spec:
              accessModes: [ ReadWriteOnce ]

      params:
        - name: source-url
          value: $(pipeline.spec.inputs.source-url)$
        - name: source-revision
          value: $(pipeline.spec.inputs.source-revision)$
```

### PR DESCRIPTION
**Pull request by martyspiewak**
Closes #247

### CODE REVIEW

File: pkg/apis/v1alpha1/cluster_deployment_template.go [hunk 1]
```diff
@@ -41,6 +66,41 @@ type ClusterDeploymentTemplateList struct {
...
+	}
+
+	if bothSet(c.Spec) || neitherSet(c.Spec) {
+		return fmt.Errorf("invalid spec: must set exactly one of spec.ObservedMatches and spec.ObservedCompletion")
+	}
+
+	return nil
+}
+
+func bothSet(spec DeploymentSpec) bool {
```
Comment (cirocosta) [inline pkg/apis/v1alpha1/cluster_deployment_template.go]: `bothSet` and `neitherSet` here end up being available (although privately) for the whole v1alpha1 package. given that _it is_ DeploymentSpec specific (given the signature of the functions), I wonder if it'd be better to instead make them DeploymentSpec methods instead 

```suggestion
func (spec DeploymentSpec) both_somethingsomething_Set() bool {
```

wdyt?

File: pkg/templates/cluster_deployment_template.go [hunk 2]
```diff
@@ -37,14 +39,110 @@ func (t clusterDeploymentTemplate) GetName() string {
 	return t.template.Name
 }
 
-func (t clusterDeploymentTemplate) GetOutput(stampedObject *unstructured.Unstructured) (*Output, error) {
-	return &Output{}, nil
+func (t clusterDeploymentTemplate) GetOutput(stampedObject *unstructured.Unstructured, templatingContext map[string]interface{}) (*Output, error) {
+	if err := t.outputReady(stampedObject); err != nil {
+		return nil, err
```
Comment (cirocosta) [inline pkg/templates/cluster_deployment_template.go:54]: not sure if intended or not - should this err be wrapped?

File: pkg/apis/v1alpha1/cluster_deployment_template.go (hunk 1 above)
Comment (waciumawanjohi) [reply pkg/apis/v1alpha1/cluster_deployment_template.go]: Good call

File: pkg/templates/cluster_deployment_template.go (hunk 2 above)
Comment (waciumawanjohi) [reply pkg/templates/cluster_deployment_template.go:54]: Better to leave it unwrapped. Carto throws typed errors if the output is not ready (e.g. the conditions on the Deployment are not met) and wrapping means that our later error handling must unwrap them further.

File: pkg/controller/deliverable/reconciler.go [hunk 3]
```diff
@@ -90,7 +91,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		case realizer.ApplyStampedObjectError:
 			r.conditionManager.AddPositive(TemplateRejectedByAPIServerCondition(typedErr))
 		case realizer.RetrieveOutputError:
-			r.conditionManager.AddPositive(MissingValueAtPathCondition(typedErr.ResourceName(), typedErr.JsonPathExpression()))
+			switch typedErr.Err.(type) {
```
Comment (emmjohnson) [inline pkg/controller/deliverable/reconciler.go:94]: I'm not sold on this, but I think we can work on this outside of this PR (my next PR for handing errors maybe?)

**Review by emmjohnson — APPROVED** [review]

File: pkg/controller/deliverable/reconciler.go (hunk 3 above)
Comment (martyspiewak) [reply pkg/controller/deliverable/reconciler.go:94]: Sounds good

**Pull request merged**